pkg/server/status_insights.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/server/authserver"
+	"github.com/cockroachdb/cockroach/pkg/server/serverpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/appstatspb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
+	"github.com/cockroachdb/cockroach/pkg/sql/clusterunique"
+	"github.com/cockroachdb/cockroach/pkg/sql/contentionpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/isql"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlstats/insights"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+	"github.com/cockroachdb/errors"
+)
+
+// StmtInsightsDetails represents the structure of details column
+// in system.statement_execution_insights table.
+// Reminder for future maintainers: keep this struct backward compatible.
+// Do not remove/rename already existing fields.
+type StmtInsightsDetails struct {
+	RowsRead    int64 `json:"rows_read"`
+	RowsWritten int64 `json:"rows_written"`
+}
+
+func (d *StmtInsightsDetails) ToJSON() (string, error) {
+	b, err := json.Marshal(d)
+	if err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+// StatementExecutionInsights requests statement insights that satisfy specified
+// parameters in request payload if any provided.
+func (s *statusServer) StatementExecutionInsights(
+	ctx context.Context, req *serverpb.StatementExecutionInsightsRequest,
+) (*serverpb.StatementExecutionInsightsResponse, error) {
+	ctx = authserver.ForwardSQLIdentityThroughRPCCalls(ctx)
+	ctx = s.AnnotateCtx(ctx)
+	if err := s.privilegeChecker.RequireViewActivityOrViewActivityRedactedPermission(ctx); err != nil {
+		return nil, err
+	}
+	// Validate that either both start and end time are defined or not.
+	if (req.StartTime != nil && req.EndTime == nil) || (req.EndTime != nil && req.StartTime == nil) {
+		return nil, errors.New("required that both StartTime and EndTime to be set")
```

Comment:
Do we have to require this or can we just omit them from the query if they're not provided?